gui/default/syncthing/core/localeService.js

```diff
@@ -78,14 +79,16 @@ angular.module('syncthing.core')
                                 // case. We compare to the lowercase version of the language
                                 // code we have as well.
                                 possibleLang = possibleLang.toLowerCase();
-                                if (possibleLang.length > lang.length) {
-                                    return possibleLang.indexOf(lang) === 0;
-                                } else {
-                                    return lang.indexOf(possibleLang) === 0;
+                                // Match "en" with "en-US" but not "fi" with "fil",
+                                // and match "en-US" with "en" but not "fil" with "fi".
+                                if (possibleLang !== lang) {
+                                    lang = lang.replace(pattern, '');
```

Comment:
This could be done once, where `lang` is first assigned in the loop.

Reply:
In line 71, where the value is assigned:

    lang = langs[i].replace(pattern, '');

Reply:
Ah, sorry I think I see what you mean.  Simply discarding the region info would match the wrong variant when there are several.  So it's OK to do it only when not finding an exact complete match.

Reply:
I'm not sure the current version is correct. It may be too simplified.

There are five possible cases.

1. The detected and available languages are the same (e.g. `zh-TW` and `zh-TW`). Use the language.
2. The detected and available language are completely different with no hyphen. Use the default locale.
3. The detected language has a hyphen but the available one does not (e.g. `zh-TW` and `zh`). Use `zh`.
4. The detected language has no hyphen but the available one does (e.g. `zh` and `zh-TW`). Use `zh-TW`.
5. Both detected and available languages have hyphens but are not the same (e.g. `zh-TW` and `zh-CN`). Use `zh-CN`.

I'm testing different possibilities right now, and for example, if we set the browser to `en-GB`, the UI actually defaults to `en-US` (or rather just `en`, which translates to `en-US`), which is wrong. The same goes for `zh-TW` which matches with `zh-CN` instead of `zh-TW`. When using just `return lang === possibleLang` with the conditional above commented out, the GUI does default to `zh-TW` instead of `zh-CN`.

English seems to be a special case, because the array looks like this:

```
[
    "en",
    "en-GB"
]
```

and the language should actually match with the second one, not just `en`.

I'm going to play around with this a little bit more.

Reply:
So it depends on the order of the available languages array?  If so, we should explicitly sort it here and then possibly make the code depend on the order.  The basic problem is that it assumes the first match is the best (`matching[0]`).  But if the sort order turns out the other way around, we might just need to take the last matching entry.

Reply:
The sort order doesn't cut it, because there can be 3+ variants of the language, so if we only rely on the order, then it won't ever match things like `zh-HK` (as it's between `zh-CN` and `zh-TW`).

The current iteration is not the greatest with the filter function basically duplicated, however it does seem to cover all edge cases. I've tested all different variants of English, all variants of Chinese, other single languages, etc. Specifically with English, `en-GB` needs to match `en-GB`, while `en` and all other `en-*` variants need to match just `en` (which in reality is `en-US` but the GUI uses simply `en` for the name).

The general idea is to look for exact matches first, and only when not found, go through the hyphenated variants, choosing the first matched. If you've got any tips how to simplify or improve the code, they will be very welcome.